### PR TITLE
Added TRACE header and "request_id" variable to streams

### PR DESCRIPTION
**Pull request by zdmytriv**
### 1. New "request_id" variable for stream

Introduced `request_id` variable 32 HEX random string for stream.

Since `request_id` became standard variable now. It can be used in logs. For example if logs setup like this.

Example. For config like this:
```
stream {
    # more stuff

    log_format stream_json_combined escape=json '{ '
      '"time": "$time_local", '
      '"trace_id": "$request_id", '
      '"remote_addr": "$proxy_protocol_addr", '
      '"protocol": "$protocol", '
      '"status": $status, '
      '"bytes_sent": $bytes_sent, '
      '"bytes_received": $bytes_received, '
      '"session_time": $session_time, '
      '"upstream_addr": "$upstream_addr" '
      ' }';

   # ... more
}
```
^ note `trace_id` gets value from `$request_id`.

Output will be
```
{ "time": "28/Mar/2020:22:46:48 +0000", "trace_id": "1b458c481644a39176ba72dc6ddbf766", "remote_addr": "", "protocol": "TCP", "status": 200, "bytes_sent": 221, "bytes_received": 144, "session_time": 83.425, "upstream_addr": "3.137.63.131:16949"  }
```
^ note value of `trace_id`.

### 2. Added `TRACE` header for stream

Introduced new `TRACE` header for stream traffic similar to proxy protocol `PROXY` header. It suppose to be used for tracing.

At the moment it will be ON once `proxy_protocol      on;`. So  it has to be being taken care in upstream.

*Format*
```
TRACE <32HEX>\r\n
```

Since it is always goes with PROXY header so first 2 lines of CONNECT request will look like this:
```
PROXY TCP4 172.17.0.1 172.17.0.2 54636 8080
TRACE 1b458c481644a39176ba72dc6ddbf766
```

*Example*
```
PROXY TCP4 172.17.0.1 172.17.0.2 54636 8080
TRACE 1b458c481644a39176ba72dc6ddbf766
CONNECT httpbin.org:443 HTTP/1.1
Host: httpbin.org:443
User-Agent: curl/7.64.1
Proxy-Connection: Keep-Alive
...
```

### 3. Added option to turn on/off "trace_protocol" header

```
stream {
    # ... more

    server {
      listen              8080   proxy_protocol;
      proxy_protocol      on;
      trace_protocol      on;     # <<<<< this is new. Note that it will effective only if proxy_protocol is on.
      proxy_pass          upstream;
    }

    # ... more
}
```

Fixes: https://github.com/opentracing/specification/issues/150

### CODE REVIEW

File: src/stream/ngx_stream_proxy_module.c [hunk 1]
```diff
@@ -907,6 +938,32 @@ ngx_stream_proxy_init_upstream(ngx_stream_session_t *s)
...
 }
 
+u_char *
+ngx_trace_protocol_write(ngx_stream_session_t *s, u_char *buf, u_char *last)
+{
+    ngx_stream_variable_value_t  *val;
+    u_char *var;
+    ngx_uint_t key;
+
+    if (last - buf < NGX_TRACE_PROTOCOL_MAX_HEADER) {
```
Comment (pvlshev) [inline src/stream/ngx_stream_proxy_module.c:948]: should also make sure that last > buf
`if (last > buf && last - buf < NGX_TRACE_PROTOCOL_MAX_HEADER)`

File: src/stream/ngx_stream_proxy_module.c [hunk 2]
```diff
@@ -907,6 +938,32 @@ ngx_stream_proxy_init_upstream(ngx_stream_session_t *s)
...
+
+    if (var == NULL) {
+        return NULL;
+    }
+
+    key = ngx_hash_strlow(var, ngx_request_id_var.data, ngx_request_id_var.len);
+    val = ngx_stream_get_variable(s, &ngx_request_id_var, key);
+
+    buf = ngx_cpymem(buf, "TRACE ", sizeof("TRACE ") - 1);
+    buf = ngx_cpymem(buf, val->data, 32);
```
Comment (pvlshev) [inline src/stream/ngx_stream_proxy_module.c:962]: lets have a constant to define that 32 is a number of bytes in trace_id, as this value also goes inside the function ngx_stream_variable_request_id in ngx_stream_variables.c 

```
#ifndef NGX_TRACE_PROTOCOL_TRACEID_LENGHT
#define NGX_TRACE_PROTOCOL_TRACEID_LENGHT (32) // "0123456789abcde0123456789abcde"
#endif //NGX_TRACE_PROTOCOL_TRACEID_LENGHT
```

File: src/stream/ngx_stream_proxy_module.c [hunk 3]
```diff
@@ -907,6 +938,32 @@ ngx_stream_proxy_init_upstream(ngx_stream_session_t *s)
...
+    var = ngx_pnalloc(s->connection->pool, ngx_request_id_var.len);
+
+    if (var == NULL) {
+        return NULL;
+    }
+
+    key = ngx_hash_strlow(var, ngx_request_id_var.data, ngx_request_id_var.len);
+    val = ngx_stream_get_variable(s, &ngx_request_id_var, key);
+
+    buf = ngx_cpymem(buf, "TRACE ", sizeof("TRACE ") - 1);
```
Comment (pvlshev) [inline src/stream/ngx_stream_proxy_module.c:961]: do not copy-paste values and change to 
```
uchar trace_id_header [] = "TRACE ";
...
buf = ngx_cpymem(buf, trace_id_header, ngx_strlen(trace_id_header));
```

File: src/stream/ngx_stream_variables.c [hunk 4]
```diff
@@ -1248,3 +1253,28 @@ ngx_stream_variables_init_vars(ngx_conf_t *cf)
...
+    }
+
+    v->valid = 1;
+    v->no_cacheable = 0;
+    v->not_found = 0;
+
+    v->len = 32;
+    v->data = id;
+
+    ngx_sprintf(id, "%08xD%08xD%08xD%08xD",
```
Comment (pvlshev) [inline src/stream/ngx_stream_variables.c:1275]: lets make sure we do not overflow at this point by using safe 
ngx_slprintf(id, id + NGX_TRACE_PROTOCOL_TRACEID_LENGHT, "%08xD%08xD%08xD%08xD", ...);

File: src/stream/ngx_stream_variables.c [hunk 5]
```diff
@@ -1248,3 +1253,28 @@ ngx_stream_variables_init_vars(ngx_conf_t *cf)
...
     return NGX_OK;
 }
+
+static ngx_int_t
+ngx_stream_variable_request_id(ngx_stream_session_t *s,
+        ngx_stream_variable_value_t *v, uintptr_t data)
+{
+    u_char  *id;
+
+    id = ngx_pnalloc(s->connection->pool, 32);
```
Comment (pvlshev) [inline src/stream/ngx_stream_variables.c:1263]: NGX_TRACE_PROTOCOL_TRACEID_LENGHT

File: src/stream/ngx_stream_variables.c [hunk 6]
```diff
@@ -1248,3 +1253,28 @@ ngx_stream_variables_init_vars(ngx_conf_t *cf)
...
+    if (id == NULL) {
+        return NGX_ERROR;
+    }
+
+    v->valid = 1;
+    v->no_cacheable = 0;
+    v->not_found = 0;
+
+    v->len = 32;
```
Comment (pvlshev) [inline src/stream/ngx_stream_variables.c:1272]: NGX_TRACE_PROTOCOL_TRACEID_LENGHT

File: src/stream/ngx_stream_proxy_module.c (hunk 1 above)
Comment (zdmytriv) [reply src/stream/ngx_stream_proxy_module.c:948]: if `last < buf` then we will get negative number here `last - buf` which is going to be less then `NGX_TRACE_PROTOCOL_MAX_HEADER`

File: src/stream/ngx_stream_variables.c (hunk 4 above)
Comment (pvlshev) [reply src/stream/ngx_stream_variables.c:1275]: @zdmytriv ^^

Comment (zdmytriv) [reply src/stream/ngx_stream_variables.c]: thanks, fixed

**Comment by pvlshev** [comment]
lgtm

**Pull request merged**